File: leetcode/20. Valid Parentheses/isValid.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = (s) => {
  const stack = [];

  if (s.length === 0) {
    return true;
  }

  if (s.length % 2 === 1) {
    return false;
  }

  for (let i = 0; i < s.length; i += 1) {
    const parentheses = s[i];
    if (parentheses === '(' || parentheses === '[' || parentheses === '{') {
      stack.push(parentheses);
    } else {
      const leftParentheses = stack.pop();
      switch (parentheses) {
        // case '(':
        // case '[':
        // case '{':
        //   stack.push(parentheses);
        //   break;
        case ')':
          if (leftParentheses !== '(') {
            return false;
          }
          break;
        case ']':
          if (leftParentheses !== '[') {
            return false;
          }
          break;
        case '}':
          if (leftParentheses !== '{') {
            return false;
          }
          break;
        default:
          return false;
      }
    }
  }

  return stack.length === 0;
};

console.log(isValid('()'));
console.log(isValid('()[]{}'));

File: leetcode/20. Valid Parentheses/isValid.cpp
#include <stack>
#include <string>

using namespace std;

/**
 * @brief 10m 54s
 * @brief 12m 47s
 * O(n)
 */
class Solution {
 public:
  // s = "()[]{}"
  bool isValid(string s) {
    stack<char> charStack;

    for (auto &&c : s) {
      if (c == '(') {
        // 1. charStack = [')']
        charStack.push(')');
      } else if (c == '[') {
        charStack.push(']');
      } else if (c == '{') {
        charStack.push('}');
      } else {
        if (charStack.empty()) {
          return false;
        }

        // 2. charStack = []
        char prevChar = charStack.top();
        charStack.pop();

        if (prevChar != c) {
          return false;
        }
      }
    }

    return charStack.empty();
  }
};
